### ISSUE TITLE
Linker flag -ObjC results in linker errors

### ISSUE THREAD

**Issue by WingedDoom** [google/j2objc#968]
Hi! I've been using J2ObjC for quite a while (thanks for a great tool!) and faced the following issue recently. I've been trying to include Firebase manually to my iOS app, and in Firebase readme it is said to add `-ObjC` linker flag.

But that makes Xcode to return errors on linking step:
```
Undefined symbols for architecture x86_64:
  "_OrgJsonJSONObject_valueToStringWithId_", referenced from:
      -[OrgJsonJSONWriter valueWithId:] in libedutatar-objc.a(JSONWriter.o)
      -[OrgJsonJSONWriter valueWithId:] in libjsoup-objc.a(JSONWriter.o)
  "_create_OrgJsonJSONException_initWithJavaLangThrowable_", referenced from:
      _OrgJsonJSONWriter_appendWithNSString_ in libedutatar-objc.a(JSONWriter.o)
      _OrgJsonJSONWriter_endWithChar_withChar_ in libedutatar-objc.a(JSONWriter.o)
      -[OrgJsonJSONWriter keyWithNSString:] in libedutatar-objc.a(JSONWriter.o)
      _OrgJsonJSONWriter_appendWithNSString_ in libjsoup-objc.a(JSONWriter.o)
      _OrgJsonJSONWriter_endWithChar_withChar_ in libjsoup-objc.a(JSONWriter.o)
      -[OrgJsonJSONWriter keyWithNSString:] in libjsoup-objc.a(JSONWriter.o)
  "_OrgJsonJSONException_initWithNSString_withJavaLangThrowable_", referenced from:
      _OrgJsonJSONPointerException_initWithNSString_withJavaLangThrowable_ in libedutatar-objc.a(JSONPointerException.o)
      _OrgJsonJSONPointerException_initWithNSString_withJavaLangThrowable_ in libjsoup-objc.a(JSONPointerException.o)
  "_create_OrgJsonJSONObject_initWithInt_", referenced from:
      _OrgJsonProperty_toJSONObjectWithJavaUtilProperties_ in libedutatar-objc.a(Property.o)
      _OrgJsonProperty_toJSONObjectWithJavaUtilProperties_ in libjsoup-objc.a(Property.o)
  "_OrgJsonJSONObject_stringToValueWithNSString_", referenced from:
      _OrgJsonXML_stringToValueWithNSString_ in libedutatar-objc.a(XML.o)
      _OrgJsonXML_stringToValueWithNSString_ in libjsoup-objc.a(XML.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

Removing the `-ObjC` flag fixes the issue, but I'm not sure if its okay to use Firebase without it.

I have set up my J2ObjC project as following:
In java I have my library package `edutatar`, and 3rd party packages `org.jsoup` and `org.json`. These packages have classes with the same names whose files get overwritten if I set up the project as in "Build Rule" J2ObjC guide. So I have created different targets for `org.*` and `edutatar`, which are static libraries with disabled ARC, and on which main app target depends. Their files get transpiled into different folders, because they have their own build rule j2objc script each. This way classes with the same names do not overwrite each other during j2objc build rule since they're in different folders.

**Comment by kpgalligan** [comment]
To be more clear, I assume you have something like
-ljre_emul -ledutatar-objc -ljsoup-objc -ObjC
in your linker args. Try moving jre_emul
-ledutatar-objc -ljsoup-objc -ljre_emul -ObjC
I have absolutely no idea if that'll work, and you're certainly better off resolving the different org.json versions, but it's an idea.
Also could try linking to a j2objc lib that doesn't include the json classes. Again, assuming that's the issue.

**Comment by tomball** [comment]
Right, there are different org.json libraries. The j2objc version uses
Android's version for compatibility, and as Kevin points out, it doesn't
have JSONObject.valueToString(). The reason the Android version of your app
may work is because Java's classloaders allow multiple versions of a class
without conflicts (because the first one on the classpath wins).

There a couple of ways you can fix this. First, does your app or any of its
dependent libraries actually need valueToString()? Its implementation
simply returns JSONWriter.valueToString(value)
<https://github.com/stleary/JSON-java/blob/master/JSONObject.java#L2324>,
so if you can instead inline that call you can drop the need for an
otherwise redundant library.

Since j2objc supports linking in subsets of the jre_emul library
<https://developers.google.com/j2objc/guides/required-link-flags#jre_libraries>,
another option is to only link in the subsets your app needs. The JSON
classes are in the android_util subset library, so unless your app uses
other classes, specifying subsets allows you to link without the default
JSON support.

To use JRE subset libraries, replace the link reference to jre_emul
(probably a -ljre_emul flag) with jre_core. You'll probably have some
unresolved symbol linker errors, so find the class in jre_sources.mk
<https://github.com/google/j2objc/blob/master/jre_emul/jre_sources.mk> to
find which subset to add.

For example, say linking fails with a missing symbol that starts with
_JavaIoCharArrayReader. Looking up java.io.CharArrayReader in jre_sources.mk
shows it's in the JAVA_PUBLIC_SOURCES_IO group
<https://github.com/google/j2objc/blob/master/jre_emul/jre_sources.mk#L674>,
so also link in jre_io to resolve it (and all other missing java.io
symbols).

The advantage of either approach is that your iOS app will be smaller,
because -ObjC tells the linker to load every class and category it finds in
every static library the app is linked with.

**Comment by WingedDoom** [comment]
Thanks for the replies!

Since I'm not a proficient Java developer, I didn't actually know J2ObjC's JRE includes `org.json` already, so I was able to remove that completely from the list of files to convert to ObjC.

But then `-ObjC` linker flag caused multiple duplicate errors from `jsoup` and our `edutatar` library.
I've read more about linker flags and realised, since `-ObjC` is only needed for Firebase-related frameworks, I can replace `-ObjC` with `-force_load /path/to/Firebase.framework/Firebase`. I did so for all the Firebase-related frameworks and it worked perfectly.

I guess `-force_load` flag could be mentioned in the J2ObjC docs somewhere near the note about `-ObjC` as an alternative for using `-ObjC`, since there are libraries requiring the use of `-ObjC`.
